x-pack/plugins/lens/public/app_plugin/save_modal_container.tsx

```diff
@@ -0,0 +1,396 @@
+      }
+    }
+    try {
+      const newInput = (await attributeService.wrapAttributes(
+        docToSave,
+        options.saveToLibrary,
+        originalInput
+      )) as LensEmbeddableInput;
+
+      if (saveProps.returnToOrigin && redirectToOrigin) {
+        // disabling the validation on app leave because the document has been saved.
+        onAppLeave?.((actions) => {
+          return actions.default();
+        });
+        redirectToOrigin({ input: newInput, isCopied: saveProps.newCopyOnSave });
+        return;
+      } else if (saveProps.dashboardId && redirectToDashboard) {
+        // disabling the validation on app leave because the document has been saved.
+        onAppLeave?.((actions) => {
+          return actions.default();
+        });
+        redirectToDashboard(newInput, saveProps.dashboardId);
+        return;
+      }
+
+      notifications.toasts.addSuccess(
+        i18n.translate('xpack.lens.app.saveVisualization.successNotificationText', {
+          defaultMessage: `Saved '{visTitle}'`,
+          values: {
+            visTitle: docToSave.title,
+          },
+        })
+      );
+
+      if (
+        attributeService.inputIsRefType(newInput) &&
+        newInput.savedObjectId !== originalSavedObjectId
+      ) {
+        chrome.recentlyAccessed.add(
+          getFullPath(newInput.savedObjectId),
+          docToSave.title,
+          newInput.savedObjectId
+        );
+
+        setIsSaveModalVisible();
+        // remove editor state so the connection is still broken after reload
+        stateTransfer.clearEditorState?.(APP_ID);
+
+        redirectTo?.(newInput.savedObjectId);
+        resolve({ isLinkedToOriginatingApp: false });
+        return;
+      }
+
+      const newDoc = {
+        ...docToSave,
+        ...newInput,
+      };
+
+      resolve({ persistedDoc: newDoc, lastKnownDoc: newDoc, isLinkedToOriginatingApp: false });
+
+      setIsSaveModalVisible();
+    } catch (e) {
+      // eslint-disable-next-line no-console
+      console.dir(e);
+      trackUiEvent('save_failed');
+      setIsSaveModalVisible();
+    }
+  });
+};
+
+export function getLastKnownDocWithoutPinnedFilters(doc?: Document) {
+  if (!doc) return undefined;
+  const [pinnedFilters, appFilters] = partition(
+    injectFilterReferences(doc.state?.filters || [], doc.references),
+    esFilters.isFilterPinned
+  );
+  return pinnedFilters?.length
+    ? {
+        ...doc,
+        state: {
+          ...doc.state,
+          filters: appFilters,
+        },
+      }
+    : doc;
+}
+
+export const getLastKnownDoc = async ({
+  initialInput,
+  attributeService,
+  data,
+  notifications,
+  chrome,
+}: {
+  initialInput?: LensEmbeddableInput;
+  attributeService: LensAttributeService;
+  data: DataPublicPluginStart;
+  notifications: NotificationsStart;
+  chrome: ChromeStart;
+}) => {
+  if (!initialInput) {
+    return;
+  }
+  return new Promise<{ doc: Document; indexPatterns: IndexPattern[] }>((resolve, reject) => {
```

Comment:
Same here. I do not think we need a `Promise` wrapper within an `async` function, for this particular case